### ISSUE TITLE
Produce RFC 8288 compliant Link headers

### ISSUE THREAD

**Issue by bblfish** [akka/akka-http#4264]
Currently, the Link header implementation produces RFC 5988 (2010) compliant Link header. But that has been superseded by RFC 8288 (2017). And as I show below, the current Akka implementation does not produce headers compliant for that later RFC, as it does not quote URLs. I discovered that by finding that the Link headers produced by Akka were not parsed correctly by http4s. See https://github.com/http4s/http4s/issues/7101

Below is the Reasoning I followed to work out what the situation is using extracts from the various RFCs.

## RFC Reasoning

In [RFC 8288](https://www.rfc-editor.org/rfc/rfc8288.html#section-3) defines the value of the Link header via the following simple ABNF:

```ABNF
Link       = #link-value
link-value = "<" URI-Reference ">" *( OWS ";" OWS link-param )
link-param = token BWS [ "=" BWS ( token / quoted-string ) ]
```

The problem we are having is with `token` and in particular the character `:` .
`token` is defined in [rfc 7230](https://www.rfc-editor.org/rfc/rfc7230#section-3.2.6) with the ABNF

```abnf
token         = 1*tchar
tchar          = "!" / "#" / "$" / "%" / "&" / "'" / "*"
                    / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
                    / DIGIT / ALPHA
                    ; any VCHAR, except delimiters
```

which as I see does not contain the `:` character. 
That seems to confirm the implementation in http4s.

The problem is that the predecessor of rfc 7230, namely [rfc 5988](https://www.rfc-editor.org/rfc/rfc5988#section-4.2)

```abnf
link-value       = "<" URI-Reference ">" *( ";" link-param )
link-param     = ( ( "rel" "=" relation-types )
                 | ( "anchor" "=" <"> URI-Reference <"> )
                 | ( "rev" "=" relation-types )
                 | ( "hreflang" "=" Language-Tag )
                 | ( "media" "=" ( MediaDesc | ( <"> MediaDesc <"> ) ) )
                 | ( "title" "=" quoted-string )
                 | ( "title*" "=" ext-value )
                 | ( "type" "=" ( media-type | quoted-mt ) )
                 | ( link-extension ) )
   ;[snip]
relation-type  = reg-rel-type | ext-rel-type
reg-rel-type   = LOALPHA *( LOALPHA | DIGIT | "." | "-" )
ext-rel-type   = URI
```
We are concerned here with `ext-rel-type`.

Now here, the ABNF is insufficiently defined it seems. `URI` is defined in 
[RFC 3986 §Appendix A](https://www.rfc-editor.org/rfc/rfc3986#appendix-A) 

```abnf
URI           = scheme ":" hier-part [ "?" query ] [ "#" fragment ]
```
without quotes, which would indicate that no quotes should be used.  But the text
in our older Link RFC [rfc5988 §5.3](https://www.rfc-editor.org/rfc/rfc5988#section-5.3)
states that:

> Note that extension relation types are REQUIRED to be absolute URIs
   in Link headers, and MUST be quoted if they contain a semicolon (";")
   or comma (",") (as these characters are used as delimiters in the
   header itself).

So there it looks like the reason they should be quoted is if those special characters
turn up, which is how Akka implemented it. Still, this shows the ABNF is incomplete, as that makes no mentions of quotes.

### Conclusion 

So it seems that the http4s implementation is correct for the more recent spec. Because
absolute URLs always contain a `:`, they cannot actually be used without quotes. Still,
there is a problem of backward compatibility now. That would be solved most simply if Akka quoted all Uris (perhaps by just also checking for `:`, or by adding a constructor to create `rel` attributes that takes an Absolute `Uri`

**Comment by johanandren** [comment]
Thanks for the detailed write up, sounds like it would be safest to always quote the URLs

**Comment by johanandren** [comment]
Or actually, maybe that is according to RFC 7230 because the value is not a valid token.

**Comment by bblfish** [comment]
Yes, so according to [Postel's Law](https://ardalis.com/postels-law-robustness-principle/) 

> Be conservative in what you do, be liberal in what you accept from others.

You should produce the headers that are most widely understandable, and that would be if they are quoted when `Uri`, but when parsing Link headers you should allow the more liberal older version (which http4s does not but should).

**Comment by johanandren** [comment]
Ok, let me rephrase that, I have something like a question hiding in there:

In your issue over at http4s, the full header is `Link: <http://localhost:8080/ldes/defaultCF/stream.acl>; rel=acl, <http://localhost:8080/ldes/defaultCF/.acl>; rel=https://www.w3.org/ns/auth/acl#accessControl` but I don't understand how you achieved that, unless you stuffed `acl, <http://localhost:8080/ldes/defaultCF/.acl>` into one `LinkParams.rel(...)`. 

That would mean that defaulting to quote the whole rel would end up with `...; rel="acl, <http://localhost:8080/ldes/defaultCF/.acl>";` which seems a bit strange to me.
